src/app-layer-dnp3.c

```diff
@@ -0,0 +1,2616 @@
+            block_size = DNP3_BLOCK_SIZE + DNP3_CRC_LEN;
+        }
+        else {
+            block_size = input_len - offset;
+        }
+
+        /* For the first block we trim the leading transport header. */
+        if (offset == 0) {
+            memcpy(*output + *output_len, input + offset + 1,
+                block_size - DNP3_CRC_LEN - 1);
+            *output_len += block_size - DNP3_CRC_LEN - 1;
+        }
+        else {
+            memcpy(*output + *output_len, input + offset,
+                block_size - DNP3_CRC_LEN);
+            *output_len += block_size - DNP3_CRC_LEN;
+        }
+        offset += block_size;
+    }
+
+    return *output;
+}
+
+static void *DNP3StateAlloc(void)
+{
+    SCEnter();
+    DNP3State *dnp3;
+
+    dnp3 = (DNP3State *)SCCalloc(1, sizeof(DNP3State));
+    if (unlikely(dnp3 == NULL)) {
+        return NULL;
+    }
+    TAILQ_INIT(&dnp3->tx_list);
+
+    dnp3->sessions = HashListTableInit(256, DNP3SessionHash,
+        DNP3SessionHashCompare, DNP3SessionFree);
+    if (unlikely(dnp3->sessions == NULL)) {
+        goto fail;
+    }
+
+    SCReturnPtr(dnp3, "void");
+fail:
+    if (dnp3 != NULL)
+        SCFree(dnp3);
+    SCReturnPtr(NULL, "void");
+}
+
+static void DNP3SetEvent(DNP3State *dnp3, uint8_t event)
+{
+    if (dnp3 && dnp3->curr) {
+        AppLayerDecoderEventsSetEventRaw(&dnp3->curr->decoder_events, event);
+        dnp3->events++;
+    }
+    else {
+        SCLogWarning(SC_ERR_ALPARSER,
+            "Fail set set event, state or txn was NULL.");
+    }
+}
+
+static DNP3Transaction *DNP3TxAlloc(DNP3State *dnp3)
+{
+    DNP3Transaction *tx = SCCalloc(1, sizeof(DNP3Transaction));
+    if (unlikely(tx == NULL)) {
+        return NULL;
+    }
+    dnp3->unreplied++;
+    dnp3->curr = tx;
+    tx->dnp3 = dnp3;
+    tx->tx_num = ++dnp3->transaction_max;
+    TAILQ_INIT(&tx->request_objects);
+    TAILQ_INIT(&tx->response_objects);
+    TAILQ_INSERT_TAIL(&dnp3->tx_list, tx, next);
+
+    /* Check for flood state. */
+    if (dnp3->unreplied > DNP3_DEFAULT_REQ_FLOOD_COUNT) {
+        DNP3SetEvent(dnp3, DNP3_DECODER_EVENT_FLOODED);
+        dnp3->flooded = 1;
+    }
+
+    return tx;
+}
+
+static DNP3Transaction *DNP3FindTransaction(DNP3State *dnp3,
+    DNP3Session *session, uint8_t seqno)
+{
+    if (dnp3->curr == NULL) {
+        return NULL;
+    }
+
+    if (dnp3->curr->session == session && dnp3->curr->app_seqno == seqno) {
+        return dnp3->curr;
+    }
+
+    DNP3Transaction *tx;
+    TAILQ_FOREACH(tx, &dnp3->tx_list, next) {
+        if (tx->session == session && tx->app_seqno == seqno) {
+            return tx;
+        }
+    }
+
+    return NULL;
+}
+
+/**
+ * \brief Calculate the length of a frame with CRCs added.
+ */
+static uint32_t DNP3CalculateUnassembledLength(uint8_t length)
+{
+    uint32_t frame_len = 0;
+    int rem;
+
+    /* Subtract the 5 bytes of the header that are included in the
+     * length. */
+    length -= DNP3_LINK_HDR_LEN;
```

Comment:
The value is verified earlier, when the check for enough data for a user data frame, but I'll add the check here, and validate the response to callers of this function.